src/AddThreepid.ts

```diff
@@ -160,22 +171,29 @@ export default class AddThreepid {
             // For separate bind, request a token directly from the IS.
             const authClient = new IdentityAuthClient();
             const identityAccessToken = (await authClient.getAccessToken()) ?? undefined;
-            return MatrixClientPeg.get()
-                .requestMsisdnToken(phoneCountry, phoneNumber, this.clientSecret, 1, undefined, identityAccessToken)
-                .then(
-                    (res) => {
-                        this.sessionId = res.sid;
-                        return res;
-                    },
-                    function (err) {
-                        if (err.errcode === "M_THREEPID_IN_USE") {
-                            err.message = _t("This phone number is already in use");
-                        } else if (err.httpStatus) {
-                            err.message = err.message + ` (Status ${err.httpStatus})`;
-                        }
-                        throw err;
-                    },
+            try {
+                const res = await MatrixClientPeg.get().requestMsisdnToken(
+                    phoneCountry,
+                    phoneNumber,
+                    this.clientSecret,
+                    1,
+                    undefined,
+                    identityAccessToken,
                 );
+                this.sessionId = res.sid;
+                return res;
+            } catch (err) {
+                if (err.errcode === "M_THREEPID_IN_USE") {
+                    throw newTranslatableError("This phone number is already in use");
```

Comment:
Diving into this more, I'm not sure it makes sense to have translatable errors 🤔

Perhaps this is more the responsibility of the UI layer than these utility methods.

To make this work, we would need to maintain the error types like `HTTPError` and `MatrixError` 
which have useful properties like `errcode` and `httpStatus` and introduce translatable versions that inherit from them.

But then we get into a bit of a weird case with `MatrixError` since you [don't even provide a message to the constructor there](https://github.com/matrix-org/matrix-js-sdk/blob/fc55c4c72a7be246130a0189fd1c9538156322f6/src/http-api/errors.ts#L50-L55). So then what is `translatedMessage` even pertaining to? It's more of a user-friendly message to print instead...

### Introducing `UserFriendlyError`

Then perhaps we instead want to have this sort of usage instead `new UserFriendlyError("This email address is already in use" , { cause: err });` which has a `translatedMessage` property just like `newTranslatableError` adds to a normal `Error`.

Does `UserFriendlyError` make sense to others?

This way people can print the error out and check the actual underlying `cause` that has all of the juicy properties. We can also make TypeScript happy and check for types (`err instanceof UserFriendlyError`) to look at the type of the underlying `cause` as well.


<details>
<summary><code>UserFriendlyError</code> downstream usage example</summary>

Imagine `checkEmailLinkClicked` emits errors like the following:

```
public async checkEmailLinkClicked() {
  // ...
  throw new UserFriendlyError("Failed to verify email address: make sure you clicked the link in the email", err);
}
```

And then let's refactor downstream usage in [`verifyEmailAddress(...)` in `src/components/views/dialogs/SetEmailDialog.tsx#L110-L139`] to use this new pattern.

```js
private verifyEmailAddress(): void {
  try {
    await this.addThreepid.checkEmailLinkClicked();
  } catch(err) {
    logger.error('Unable to verify email address', err)

    let underlyingError = err;
    if(err instanceof UserFriendlyError) {
      underlyingError = err.cause;
    }

    if(underlyingError instanceof MatrixError && underlyingError.errcode === 'M_THREEPID_AUTH_FAILED') {
      const message =
          _t("Unable to verify email address.") +
          " " +
          _t(
              "Please check your email and click on the link it contains. Once this is done, click continue.",
          );
      Modal.createDialog(QuestionDialog, {
          title: _t("Verification Pending"),
          description: message,
          button: _t("Continue"),
          onFinished: this.onEmailDialogFinished,
      });
    } else {
      Modal.createDialog(ErrorDialog, {
          title: _t("Unable to verify email address."),
          description:
            (err instanceof UserFriendlyError && err.translatedMessage) ||
            (err instanceof Error && err.message) ||
            _t("Operation failed"),
      });
    }
  }
}
```

Reply:
This is being introduced via https://github.com/matrix-org/matrix-react-sdk/pull/10440